### ISSUE TITLE
DynamicViT: Efﬁcient Vision Transformers with Dynamic Token Sparsiﬁcation

### ISSUE THREAD

**Issue by yskim0** [One-Day-One-Paper/One-Day-One-Paper#56]
## Abstract

> the final prediction in vision transformers is **only based on a subset of most informative tokens**, which is sufﬁcient for accurate image recognition.
> 
- token learner와 비슷. 필요한 토큰만 쓰겠다는 방향

> Based on this observation, we propose a dynamic token sparsiﬁcation framework to prune redundant tokens progressively and dynamically based on the input. Speciﬁcally, **we devise a lightweight prediction module to estimate the importance score of each token given the current features.** The module is **added to different layers** to prune redundant tokens hierarchically.
> 
- tokenlearner 처럼 블락 뒤에 적용할 수 있는 모듈 고안
    - 이 모듈은 현재 상황에서 토큰들의 Importance score를 구할 수 있도록 함.

> To optimize the prediction module in an **end-to-end manner**, we propose an attention masking strategy to differentiably prune a token by blocking its interactions with other tokens.
> 
- masking strategy

> By hierarchically pruning 66% of the input tokens, our method greatly reduces 31% ∼ 37% FLOPs and improves the throughput by over 40% while the drop of accuracy is within 0.5% for various vision transformers.

## Introduction

- CNN-type network → prune the filters that are of **less important**
    
    > That is, **we can prune the tokens of less importance in the input instance, given the fact that many tokens contribute very little to the ﬁnal prediction.**
    > 
    
    > This is **only possible for the transformer-like models** where the self-attention module can take the token sequence of variable length as input, and the unstructured pruned input will not affect the self-attention module, while dropping a certain part of the pixels can not really accelerate the convolution operation since the unstructured neighborhood used by convolution would make it difﬁcult to accelerate through parallel computing.
    > 
- Token 줄이는 방향 ⇒ Dynamic Token Sparsification
    
    
![image](https://user-images.githubusercontent.com/48315997/151301980-a939ce94-0229-4760-bf51-be09be0f97ce.png)

    
    - CNN → (a) Structural Downsampling
    - Ours → (b) Dynamic Token Sparsification : **unstructured and data-dependent**
        
        > we hope to explore the unstructured and data-dependent downsampling strategy for vision transformers to further leverage the advantages of self-attention (our experiments alsno show unstructured sparsiﬁcation can lead to better performance for vision transformers compared to structural downsampling).
        > 

> we propose to employ a lightweight prediction module to determine which tokens to be pruned in a dynamic way, dubbed as **DynamicViT**.
> 
- module → prediction module produces **a customized binary decision mask** to decide which tokens are uninformative and need to be abandoned.
    
    > This module is added to multiple layers of the vision transformer, such that the sparsiﬁcation can be performed in a hierarchical way as we gradually increase the amount of pruned tokens after each prediction module. Once a token is pruned after a certain layer, it will not be ever used in the feed-forward procedure
    > 
- prediction module이 더해져서 End-to-end 를 하게 만드려면 2가지 전략이 필요함
    1. `Gumbel-Softmax`
    2. `how to apply this learned binary decision mask to prune the unnecessary tokens.`
        - (1) the number of zero elements in the binary decision mask is different for each instance → since zero vectors will still affect the calculation of the attention matrix.
        - (2) directly eliminating the uninformative tokens for each input instance during training → will make parallel computing impossible.
        
        ⇒ 
        
        > Therefore, we propose a strategy called **attention masking** where we drop the connection from abandoned tokens to all other tokens in the attention matrix based on the binary decision mask. By doing so, we can overcome the difﬁculties described above. We also modify the original training objective of the vision transformer by adding a term to constrain the proportion of pruned tokens after a certain layer.
        > 
        
        > During the **inference phase**, we can directly abandon a ﬁxed amount of tokens after certain layers for each input instance as we no longer need to consider whether the operation is differentiable, and this will greatly accelerate the inference.
        > 
- DeiT, LV-ViT 백본에 사용
    
    > The experimental results demonstrate the competitive trade-off between speed and accuracy. In particular, by hierarchically pruning 66% of the input tokens, we can greatly reduce 31% ∼ 37% GFLOPs and improve the throughput by over 40% while the drop of accuracy is within 0.5% for all different vision transformers.
    > 
    

## Dynamic Vision Transformers

<aside>
💡 **Overview**

</aside>

![image](https://user-images.githubusercontent.com/48315997/151302015-d57381f4-22fb-4f7d-833b-eb8bde0098f8.png)

- noram vision transformer as the backbone + several prediction modules
- prediction module 의 역할 : generating the probabilities of dropping/keeping the tokens
    
    > The token sparsiﬁcation is performed hierarchically through the whole network at certain locations. For example, given a 12-layer transformer, we can conduct token sparsiﬁcation before the 4th, 7th, and 10th blocks.
    > 

> During training, **the prediction modules and the backbone network can be optimized in an end-to-end manner** thanks to our newly devised **attention masking strategy.**
> 

> During inference, **we only need to select the most informative tokens** according to a predeﬁned pruning ratio and the scores computed by the prediction modules.
> 

<aside>
💡 **Hierarchical Token Sparsification with Prediction Modules**

</aside>

- decision mask : to indicate whether to drop or keep each token
    - $\hat D \in \{0,1\}^N, N=HW$ (N= #of patch embedding)
- process
    - initialize all elements in the decision mask to 1 and update the mask progressively
    - (1) local feature first project the tokens using an MLP
        - $z^{local} = MLP(x) \in \mathbb{R}^{N\times C'}$
        - $C'= C/2$
    - (2) global feature : aggregate the information all the existing tokens by using average pooling
        - $z^{global} = Agg(MLP(x), \hat D)\in \mathbb{R}^{C'}$
    - (1) local feature → encodes the information of a certain token
    (2) global feature → contains the context of the whole image
    - combine (1), (2) to obtain local-global embeddings and feed them to another MLP to predict probabilities to drop/keep the tokens
        
        $$
        z_i = [z_i^{local}, z_i^{global}], 1 \leq i \leq N, \\
        \pi = Softmax(MLP(z))\in \mathbb{R}^{N\times 2}
        $$
        
        
![image](https://user-images.githubusercontent.com/48315997/151302133-f296ebfa-6650-4197-b5cc-b71d0ed20c23.png)
        

<aside>
💡 **End-to-end Optimization with Attention Masking**

</aside>

- sampling from $\pi$ to get binary decision mask D is non-differentiable ⇒ impedes the end-to-end training
    - Solution) Gumbel-Softmax
        - $D = GumbelSoftmax(\pi)_{*,1} \in \{0,1\}^N$
        - use the index 1 because D represents the mask of the `kept` tokens
- how to prune the tokens during training
    - approach
    1) simply discarding the tokens where D i = 0 would result in a non-uniform number of tokens for samples within a batch ⇒ hard to paralleize the computataion
        
        2) merely zero-out the tokens to be dropped using the binary mask D is not feasible, because in the calculation of self-attention matrix
        
        the zeroed tokens will still infulence other tokens through the Softmax
        
    - To this end, `attention masking` which can totally eliminate the effects of the dropped tokens
        
       
![image](https://user-images.githubusercontent.com/48315997/151302077-8eb70d15-2024-42e0-b960-30d0f543da74.png)

        
        자기는 무조건 뽑히게 함. 
        
        P → attention value
        
        tilde_A → attention score. 분모에 only kept tokens만 생각하고 attention matrix가 계산되어짐.
